File: ios/SonyPtpProperties.swift
import Foundation

/// Availability reported by the `IsEnabled` field of an `SDIExtDevicePropInfo` record.
///
/// A property can be supported by the body yet invalid in the camera's current mode.
/// Callers need to tell that apart from "this body does not have this property" because
/// the two conditions need different UI and different error copy.
enum SonyPropertyAvailability {
  /// `IsEnabled = 0x00`. Supported but invalid in the current mode.
  case disabled
  /// `IsEnabled = 0x01`. Readable, and writable when `writable` is also true.
  case enabled
  /// `IsEnabled = 0x02`. Indication only; the value must not be written.
  case displayOnly
}

/// One value decoded from a PTP payload.
///
/// The associated types mirror the PTP datatype space: every integer datatype decodes to
/// `.integer`, `STR` to `.text`, array datatypes to `.array`, and the 128-bit datatypes
/// to `.opaque` (correctly sized, coarsely represented).
enum SonyPropertyValue: Equatable {
  case integer(Int64)
  case text(String)
  case array([SonyPropertyValue])
  case opaque(String)

  var integerValue: Int64? {
    if case let .integer(value) = self { return value }
    return nil
  }
}

struct SonyPropertyRange: Equatable {
  let minimum: SonyPropertyValue
  let maximum: SonyPropertyValue
  let step: SonyPropertyValue
}

/// One decoded device property from `SDIO_GetAllExtDevicePropInfo` (0x9209).
struct SonyDeviceProperty: Equatable {
  let code: UInt16
  let dataType: UInt16
  let writable: Bool
  let availability: SonyPropertyAvailability
  let defaultValue: SonyPropertyValue
  let currentValue: SonyPropertyValue
  let range: SonyPropertyRange?
  let settableValues: [SonyPropertyValue]
  let supportedValues: [SonyPropertyValue]

  var enabled: Bool { availability != .disabled }

  static func == (lhs: SonyDeviceProperty, rhs: SonyDeviceProperty) -> Bool {
    lhs.code == rhs.code && lhs.dataType == rhs.dataType && lhs.writable == rhs.writable
      && lhs.availability == rhs.availability && lhs.currentValue == rhs.currentValue
  }
}

struct SonyPtpParseFailure: LocalizedError {
  let message: String
  var errorDescription: String? { message }
}

/// Parser for Sony's `SDIDevicePropInfo Dataset Array`.
///
/// Array layout:
/// ```
/// Num of Elements            UINT64
/// SDIExtDevicePropInfo[0 .. Num of Elements - 1]
/// ```
///
/// Each `SDIExtDevicePropInfo` record:
/// ```
/// Device Property Code       UINT16
/// DataType                   UINT16
/// GetSet                     UINT8    0 = read only, 1 = the initiator can set the value
/// IsEnabled                  UINT8    0 = false, 1 = true, 2 = display only
/// Factory Default Value      variable, sized by DataType
/// Current Value              variable, sized by DataType
/// Form Flag                  UINT8    0 = none, 1 = range, 2 = enumeration
///   range form:              minimum, maximum, step   each sized by DataType
///   enumeration form:        UINT16 count + values    (settable list)
///                            UINT16 count + values    (supported list)
/// ```
///
/// The previous implementation searched the payload for a two-byte property code and
/// assumed a fixed offset to the current value. That could match bytes inside an earlier
/// record's enumeration list and return an unrelated value, and it could not enumerate.
/// This parser walks records sequentially so every offset is derived from the record that
/// precedes it.
enum SonyPtpProperties {
  private static let maxElements: Int64 = 4_096
  private static let formRange: UInt8 = 0x01
  private static let formEnumeration: UInt8 = 0x02

  static func parseAll(_ data: Data) -> [SonyDeviceProperty] {
    let reader = SonyPtpReader(data)
    var properties: [SonyDeviceProperty] = []
    do {
      let declared = try reader.u64()
      let count = (declared < 0 || declared > maxElements) ? maxElements : declared
      var index: Int64 = 0
      while index < count, reader.remaining > 0 {
        properties.append(try parseRecord(reader))
        index += 1
      }
    } catch {
      // A truncated payload still carries every record that preceded the truncation.
      // Returning them is strictly better than discarding a valid prefix.
    }
    return properties
  }

  static func find(_ properties: [SonyDeviceProperty], code: UInt16) -> SonyDeviceProperty? {
    properties.first { $0.code == code }
  }

  /// Convenience for the readiness properties the transport polls (D213, D215, D221).
  static func scalar(_ properties: [SonyDeviceProperty], code: UInt16) -> Int64? {
    find(properties, code: code)?.currentValue.integerValue
  }

  private static func parseRecord(_ reader: SonyPtpReader) throws -> SonyDeviceProperty {
    let code = try reader.u16()
    let dataType = try reader.u16()
    let getSet = try reader.u8()
    let isEnabled = try reader.u8()
    let defaultValue = try reader.value(dataType)
    let currentValue = try reader.value(dataType)
    let formFlag = try reader.u8()

    var range: SonyPropertyRange?
    var settableValues: [SonyPropertyValue] = []
    var supportedValues: [SonyPropertyValue] = []

    switch formFlag {
    case formRange:
      range = SonyPropertyRange(
        minimum: try reader.value(dataType),
        maximum: try reader.value(dataType),
        step: try reader.value(dataType)
      )
    case formEnumeration:
      settableValues = try reader.valueList(dataType)
      // Sony emits a second list. It is absent on some firmware, so a clean end of
      // payload here is a complete record rather than a parse failure.
      supportedValues = reader.remaining >= 2 ? try reader.valueList(dataType) : []
    default:
      break
    }

    let availability: SonyPropertyAvailability
    switch isEnabled {
    case 0: availability = .disabled
    case 2: availability = .displayOnly
    default: availability = .enabled
    }

    return SonyDeviceProperty(
      code: code,
      dataType: dataType,
      writable: getSet == 1,
      availability: availability,
      defaultValue: defaultValue,
      currentValue: currentValue,
      range: range,
      settableValues: settableValues,
      supportedValues: supportedValues
    )
  }
}

/// Little-endian cursor over a PTP payload.
///
/// Every read is bounds-checked. Advancing the cursor by the true size of each field is
/// what makes sequential record parsing possible, so an unreadable field must stop the
/// walk rather than silently return a zero and desynchronise every following record.
final class SonyPtpReader {
  private let data: Data
  private var offset: Int

  private static let maxElements: Int64 = 65_536
  private static let arrayFlag: UInt16 = 0x4000

  private enum DataType {
    static let int8: UInt16 = 0x0001
    static let uint8: UInt16 = 0x0002
    static let int16: UInt16 = 0x0003
    static let uint16: UInt16 = 0x0004
    static let int32: UInt16 = 0x0005
    static let uint32: UInt16 = 0x0006
    static let int64: UInt16 = 0x0007
    static let uint64: UInt16 = 0x0008
    static let int128: UInt16 = 0x0009
    static let uint128: UInt16 = 0x000A
    static let string: UInt16 = 0xFFFF
  }

  init(_ data: Data, offset: Int = 0) {
    self.data = data
    self.offset = offset
  }

  var remaining: Int { data.count - offset }

  func u8() throws -> UInt8 {
    let bytes = try take(1)
    return bytes[0]
  }

  func u16() throws -> UInt16 {
    let bytes = try take(2)
    return UInt16(bytes[0]) | (UInt16(bytes[1]) << 8)
  }

  func u32() throws -> Int64 {
    let bytes = try take(4)
    var result: Int64 = 0
    for index in stride(from: 3, through: 0, by: -1) {
      result = (result << 8) | Int64(bytes[index])
    }
    return result
  }

  func u64() throws -> Int64 {
    let bytes = try take(8)
    var result: UInt64 = 0
    for index in stride(from: 7, through: 0, by: -1) {
      result = (result << 8) | UInt64(bytes[index])
    }
    return Int64(bitPattern: result)
  }

  /// Reads one value of `dataType` and advances past it.
  ///
  /// An unknown datatype of known width is recoverable because the cursor still advances
  /// correctly; an unknown width is not, and throws.
  func value(_ dataType: UInt16) throws -> SonyPropertyValue {
    switch dataType {
    case DataType.int8: return .integer(signed(Int64(try u8()), bits: 8))
    case DataType.uint8: return .integer(Int64(try u8()))
    case DataType.int16: return .integer(signed(Int64(try u16()), bits: 16))
    case DataType.uint16: return .integer(Int64(try u16()))
    case DataType.int32: return .integer(signed(try u32(), bits: 32))
    case DataType.uint32: return .integer(try u32())
    // A UINT64 above Int64.max is returned as its two's-complement value. No Sony device
    // property uses that range; the cursor position is what matters here.
    case DataType.int64, DataType.uint64: return .integer(try u64())
    case DataType.int128, DataType.uint128: return .opaque(try hex(16))
    case DataType.string: return .text(try string())
    default:
      guard dataType & Self.arrayFlag != 0 else {
        throw SonyPtpParseFailure(message: String(format: "Unknown PTP datatype 0x%04X.", dataType))
      }
      return .array(try array(elementType: dataType & ~Self.arrayFlag))
    }
  }

  /// Reads a `UINT16` count followed by that many `dataType`-sized values.
  func valueList(_ dataType: UInt16) throws -> [SonyPropertyValue] {
    let count = try u16()
    guard Int64(count) <= Self.maxElements else {
      throw SonyPtpParseFailure(message: "Enumeration count \(count) is implausible.")
    }
    return try (0..<count).map { _ in try value(dataType) }
  }

  /// PTP string: `UINT8` character count, then that many UTF-16LE units including the null.
  private func string() throws -> String {
    let characters = Int(try u8())
    if characters == 0 { return "" }
    let bytes = try take(characters * 2)
    let decoded = String(data: Data(bytes), encoding: .utf16LittleEndian) ?? ""
    return decoded.trimmingCharacters(in: CharacterSet(charactersIn: "\0 "))
  }

  /// PTP array: `UINT32` element count, then that many elements of the element datatype.
  private func array(elementType: UInt16) throws -> [SonyPropertyValue] {
    let count = try u32()
    guard count >= 0, count <= Self.maxElements else {
      throw SonyPtpParseFailure(message: "Array length \(count) is implausible.")
    }
    return try (0..<count).map { _ in try value(elementType) }
  }

  private func hex(_ count: Int) throws -> String {
    let bytes = try take(count)
    return bytes.reversed().map { String(format: "%02X", $0) }.joined()
  }

  private func signed(_ raw: Int64, bits: Int) -> Int64 {
    let signBit: Int64 = 1 << (bits - 1)
    return (raw & signBit) != 0 ? raw - (1 << bits) : raw
  }

  private func take(_ size: Int) throws -> [UInt8] {
    guard size >= 0, offset + size <= data.count else {
      throw SonyPtpParseFailure(message: "PTP payload ended after \(offset) of \(data.count) bytes.")
    }
    let start = data.index(data.startIndex, offsetBy: offset)
    let end = data.index(start, offsetBy: size)
    offset += size
    return [UInt8](data[start..<end])
  }
}

File: android/src/main/java/expo/modules/sonycamera/SonyPtpProperties.kt
package expo.modules.sonycamera

/**
 * Parser for Sony's `SDIDevicePropInfo Dataset Array`, the payload returned by
 * `SDIO_GetAllExtDevicePropInfo` (0x9209).
 *
 * Array layout:
 * ```
 * Num of Elements            UINT64
 * SDIExtDevicePropInfo[0 .. Num of Elements - 1]
 * ```
 *
 * Each `SDIExtDevicePropInfo` record:
 * ```
 * Device Property Code       UINT16
 * DataType                   UINT16
 * GetSet                     UINT8    0 = read only, 1 = the initiator can set the value
 * IsEnabled                  UINT8    0 = false, 1 = true, 2 = display only
 * Factory Default Value      variable, sized by DataType
 * Current Value              variable, sized by DataType
 * Form Flag                  UINT8    0 = none, 1 = range, 2 = enumeration
 *   range form:              minimum, maximum, step   each sized by DataType
 *   enumeration form:        UINT16 count + values    (settable list)
 *                            UINT16 count + values    (supported list)
 * ```
 *
 * The previous implementation searched the payload for a two-byte property code and
 * assumed a fixed offset to the current value. That could match bytes inside an earlier
 * record's enumeration list and return an unrelated value, and it could not enumerate.
 * This parser walks records sequentially so every offset is derived from the record that
 * precedes it.
 */
internal object SonyPtpProperties {
  /** Sony reports the element count as a UINT64; no camera returns anything near this. */
  private const val MAX_ELEMENTS = 4_096

  /** Guards against a corrupt count field turning into a multi-gigabyte allocation. */
  private const val MAX_ARRAY_ELEMENTS = 65_536

  fun parseAll(data: ByteArray): List<SonyDeviceProperty> {
    val reader = SonyPtpReader(data)
    val properties = ArrayList<SonyDeviceProperty>()
    return try {
      val declared = reader.u64()
      val count = if (declared < 0 || declared > MAX_ELEMENTS) MAX_ELEMENTS.toLong() else declared
      var index = 0L
      while (index < count && reader.remaining() > 0) {
        properties.add(parseRecord(reader))
        index += 1
      }
      properties
    } catch (error: SonyPtpParseException) {
      // A truncated payload still carries every record that preceded the truncation.
      // Returning them is strictly better than discarding a valid prefix.
      properties
    }
  }

  fun find(properties: List<SonyDeviceProperty>, code: Int): SonyDeviceProperty? =
    properties.firstOrNull { it.code == code }

  /** Convenience for the readiness properties the transport polls (D213, D215, D221). */
  fun scalar(properties: List<SonyDeviceProperty>, code: Int): Long? =
    (find(properties, code)?.currentValue as? Long)

  private fun parseRecord(reader: SonyPtpReader): SonyDeviceProperty {
    val code = reader.u16()
    val dataType = reader.u16()
    val getSet = reader.u8()
    val isEnabled = reader.u8()
    val defaultValue = reader.value(dataType)
    val currentValue = reader.value(dataType)
    val formFlag = reader.u8()

    var range: SonyPropertyRange? = null
    var settableValues = emptyList<Any?>()
    var supportedValues = emptyList<Any?>()

    when (formFlag) {
      FORM_RANGE -> {
        range = SonyPropertyRange(
          minimum = reader.value(dataType),
          maximum = reader.value(dataType),
          step = reader.value(dataType),
        )
      }
      FORM_ENUMERATION -> {
        settableValues = reader.valueList(dataType)
        // Sony emits a second list. It is absent on some firmware, so a clean end of
        // payload here is a complete record rather than a parse failure.
        supportedValues = if (reader.remaining() >= 2) reader.valueList(dataType) else emptyList()
      }
    }

    return SonyDeviceProperty(
      code = code,
      dataType = dataType,
      writable = getSet == 1,
      availability = when (isEnabled) {
        0 -> SonyPropertyAvailability.DISABLED
        2 -> SonyPropertyAvailability.DISPLAY_ONLY
        else -> SonyPropertyAvailability.ENABLED
      },
      defaultValue = defaultValue,
      currentValue = currentValue,
      range = range,
      settableValues = settableValues,
      supportedValues = supportedValues,
    )
  }

  private const val FORM_RANGE = 0x01
  private const val FORM_ENUMERATION = 0x02

  /** Reads a `UINT16` count followed by that many DataType-sized values. */
  private fun SonyPtpReader.valueList(dataType: Int): List<Any?> {
    val count = u16()
    if (count > MAX_ARRAY_ELEMENTS) throw SonyPtpParseException("Enumeration count $count is implausible.")
    return (0 until count).map { value(dataType) }
  }
}

internal enum class SonyPropertyAvailability {
  /** `IsEnabled = 0x00`. Supported by the body but invalid in the current mode. */
  DISABLED,

  /** `IsEnabled = 0x01`. Readable, and writable when `writable` is also true. */
  ENABLED,

  /** `IsEnabled = 0x02`. Indication only; the value must not be written. */
  DISPLAY_ONLY,
}

internal data class SonyPropertyRange(val minimum: Any?, val maximum: Any?, val step: Any?)

/**
 * One decoded device property.
 *
 * [currentValue] is a [Long] for every integer datatype, a [String] for `STR`, a
 * `List<Any?>` for array datatypes, and a hex [String] for the 128-bit datatypes.
 */
internal data class SonyDeviceProperty(
  val code: Int,
  val dataType: Int,
  val writable: Boolean,
  val availability: SonyPropertyAvailability,
  val defaultValue: Any?,
  val currentValue: Any?,
  val range: SonyPropertyRange?,
  val settableValues: List<Any?>,
  val supportedValues: List<Any?>,
) {
  val enabled: Boolean get() = availability != SonyPropertyAvailability.DISABLED
}

internal class SonyPtpParseException(message: String) : Exception(message)

/**
 * Little-endian cursor over a PTP payload.
 *
 * Every read is bounds-checked. Advancing the cursor by the true size of each field is
 * what makes sequential record parsing possible, so an unreadable field must stop the
 * walk rather than silently return a zero and desynchronise every following record.
 */
internal class SonyPtpReader(private val data: ByteArray, private var offset: Int = 0) {
  fun remaining(): Int = data.size - offset

  fun u8(): Int = read(1) { data[it].toInt() and 0xFF }

  fun u16(): Int = read(2) {
    (data[it].toInt() and 0xFF) or ((data[it + 1].toInt() and 0xFF) shl 8)
  }

  fun u32(): Long = read(4) {
    var result = 0L
    for (index in 3 downTo 0) result = (result shl 8) or (data[it + index].toLong() and 0xFF)
    result
  }

  fun u64(): Long = read(8) {
    var result = 0L
    for (index in 7 downTo 0) result = (result shl 8) or (data[it + index].toLong() and 0xFF)
    result
  }

  /**
   * Reads one value of [dataType] and advances past it.
   *
   * Returns `null` for datatypes this parser does not decode, but only after advancing
   * the cursor by the correct number of bytes. An unknown datatype of known width is
   * recoverable; an unknown width is not.
   */
  fun value(dataType: Int): Any? = when (dataType) {
    DATATYPE_INT8 -> signed(u8().toLong(), 8)
    DATATYPE_UINT8 -> u8().toLong()
    DATATYPE_INT16 -> signed(u16().toLong(), 16)
    DATATYPE_UINT16 -> u16().toLong()
    DATATYPE_INT32 -> signed(u32(), 32)
    DATATYPE_UINT32 -> u32()
    // A UINT64 above Long.MAX_VALUE is returned as its two's-complement Long. No Sony
    // device property uses that range; the cursor position is what matters here.
    DATATYPE_INT64, DATATYPE_UINT64 -> u64()
    DATATYPE_INT128, DATATYPE_UINT128 -> hex(16)
    DATATYPE_STRING -> string()
    else ->
      if (dataType and ARRAY_FLAG != 0) array(dataType and ARRAY_FLAG.inv())
      else throw SonyPtpParseException("Unknown PTP datatype 0x%04X.".format(dataType))
  }

  /** PTP string: `UINT8` character count, then that many UTF-16LE units including the null. */
  private fun string(): String {
    val characters = u8()
    if (characters == 0) return ""
    val bytes = read(characters * 2) { start -> data.copyOfRange(start, start + characters * 2) }
    return String(bytes, Charsets.UTF_16LE).trimEnd('\u0000', ' ')
  }

  /** PTP array: `UINT32` element count, then that many elements of the element datatype. */
  private fun array(elementType: Int): List<Any?> {
    val count = u32()
    if (count < 0 || count > MAX_ELEMENTS) throw SonyPtpParseException("Array length $count is implausible.")
    return (0 until count).map { value(elementType) }
  }

  private fun hex(bytes: Int): String = read(bytes) { start ->
    (0 until bytes).joinToString("") { "%02X".format(data[start + bytes - 1 - it]) }
  }

  private fun signed(raw: Long, bits: Int): Long {
    val signBit = 1L shl (bits - 1)
    return if (raw and signBit != 0L) raw - (1L shl bits) else raw
  }

  private fun <T> read(size: Int, decode: (Int) -> T): T {
    if (size < 0 || offset + size > data.size) {
      throw SonyPtpParseException("PTP payload ended after $offset of ${data.size} bytes.")
    }
    val start = offset
    offset += size
    return decode(start)
  }

  private companion object {
    const val MAX_ELEMENTS = 65_536L
    const val ARRAY_FLAG = 0x4000

    const val DATATYPE_INT8 = 0x0001
    const val DATATYPE_UINT8 = 0x0002
    const val DATATYPE_INT16 = 0x0003
    const val DATATYPE_UINT16 = 0x0004
    const val DATATYPE_INT32 = 0x0005
    const val DATATYPE_UINT32 = 0x0006
    const val DATATYPE_INT64 = 0x0007
    const val DATATYPE_UINT64 = 0x0008
    const val DATATYPE_INT128 = 0x0009
    const val DATATYPE_UINT128 = 0x000A
    const val DATATYPE_STRING = 0xFFFF
  }
}
